extensions/application/init.lua

```diff
@@ -48,4 +48,145 @@ function application:activate(_allWindows)
+--- -- by window title
+--- hs.application'bash':name() --> Terminal
+function application.find(hint,exact)
+  if hint==nil then return end
+  local typ=type(hint)
+  if typ=='number' then return application.applicationForPID(hint)
+  elseif typ~='string' then error('hint must be a number or string',2) end
+  local r=application.applicationsForBundleID(hint)
+  if #r>0 then return tunpack(r) end
+  local apps=application.runningApplications()
+
+  if exact then for _,a in ipairs(apps) do if a:name()==hint then r[#r+1]=a end end
+  else for _,a in ipairs(apps) do if a:name():lower():find(hint:lower()) then r[#r+1]=a end end end
+  tsort(r,function(a,b)return a:kind()>b:kind()end) -- gui apps first
+  if exact or #r>0 then return tunpack(r) end
+
+  r=tpack(window.find(hint))
+  local rs={} for _,w in ipairs(r) do rs[w:application()]=true end -- :toSet
+  for a in pairs(rs) do r[#r+1]=a end -- and back, no dupes
+  if #r>0 then return tunpack(r) end
+end
+
+--- hs.application:findWindow(titlePattern) -> hs.window object(s)
+--- Method
+--- Finds windows from this application
+---
+--- Parameters:
+---  * titlePattern - a string pattern that matches (via `string.find`) the window title(s) as per `hs.window:title()` (for convenience, the matching will be done on lowercased strings)
+---
+--- Returns:
+---  * one or more hs.window objects belonging to this application that match the supplied search criterion, or `nil` if none found
+
+function application:findWindow(hint)
+  return window.find(hint,false,self:allWindows())
+end
+
+--- hs.application:getWindow(title) -> hs.window object
+--- Method
+--- Gets a specific window from this application
+---
+--- Parameters:
+---  * title - the desired window's title string as per `hs.window:title()`
+---
+--- Returns:
+---  * the desired hs.window object belonging to this application, or `nil` if not found
+function application:getWindow(hint)
+  return tpack(window.find(hint,true,self:allWindows()),nil)[1]
+end
+
+--- hs.application.open(app[, wait]) -> hs.application object
+--- Constructor
+--- Launches an application, or activates it if it's already running
+---
+--- Parameters:
+---  * app - a string describing the application to open; it can be:
+---    - the application's name as per `hs.application:name()`
+---    - the full path to an application on disk (including the `.app` suffix)
+---    - the application's bundle ID as per `hs.application:bundleID()`
+---  * wait - (optional) the maximum number of seconds to wait for the app to get into "launching" state, if not already running; if omitted, defaults to 0;
+---           if the app takes longer than this to launch, this function will return `nil`, but the app will still launch
+---
+--- Returns:
+---  * the `hs.application` object for the launched or activated application; `nil` if not found
+---
+--- Notes:
+---  * the `wait` parameter will *halt the entire script* in order to return the application object "synchronously"; only use it if you
```

Comment:
Rather than "halt the entire script", maybe this should say "block all Hammerspoon activity", since timer.usleep() will block the entire application.